File: src/HolzShots/Net/UploadHelper.vb
Imports HolzShots.Interop
Imports HolzShots.UI.Dialogs
Imports HolzShots.Windows.Forms

Namespace Net
    Friend Class UploadHelper
        Private Sub New()
        End Sub

        Friend Shared Function GetUploadReporterForCurrentSettingsContext(ByVal settingsContext As HSSettings, ByVal parentWindow As IWin32Window) As IUploadProgressReporter

#If DEBUG Then
            Dim reporters = New List(Of IUploadProgressReporter)(3) From {
                New ConsoleProgressReporter()
            }
#Else
            Dim reporters = New List(Of IUploadProgressReporter)(2)
#End If

            If settingsContext.ShowUploadProgress Then
                reporters.Add(New StatusToaster())
            End If

            If parentWindow.GetHandle() <> IntPtr.Zero Then
                reporters.Add(New TaskbarItemProgressReporter(parentWindow.Handle))
            End If

            Return New AggregateProgressReporter(reporters)
        End Function

        Friend Shared Sub InvokeUploadFinishedUi(result As UploadResult, settingsContext As HSSettings)
            Debug.Assert(result IsNot Nothing)
            Debug.Assert(Not String.IsNullOrWhiteSpace(result.Url))

            Select Case settingsContext.ActionAfterUpload
                Case UploadHandlingAction.Flyout
                    Dim lv As New UploadResultForm(result, settingsContext)
                    lv.Show()

                Case UploadHandlingAction.CopyToClipboard

                    If ClipboardEx.SetText(result.Url) Then
                        HumanInterop.ShowCopyConfirmation(result.Url)
                    ElseIf settingsContext.ShowCopyConfirmation Then
                        HumanInterop.CopyingFailed(result.Url)
                    End If

                Case UploadHandlingAction.None ' Intentionally do nothing
                Case Else ' Intentionally do nothing
            End Select
        End Sub
    End Class
End Namespace

File: src/HolzShots.Windows/Net/UploadHelper.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.Windows.Forms;
using HolzShots.Net;
using HolzShots.Windows.Forms;

namespace HolzShots.Windows.Net;

public static class UploadHelper
{
    public static ITransferProgressReporter GetUploadReporterForCurrentSettingsContext(HSSettings settingsContext, IWin32Window parentWindow)
    {
        var reporters = new List<ITransferProgressReporter>(4);
#if DEBUG
        reporters.Add(new ConsoleProgressReporter());
#endif
        if (settingsContext.ShowUploadProgress)
            reporters.Add(new UploadStatusFlyoutForm());

        if (parentWindow.GetHandle() != nint.Zero)
            reporters.Add(new TaskbarItemProgressReporter(parentWindow.Handle));

        return new AggregateProgressReporter(reporters);
    }

    public static void InvokeUploadFinishedUI(UploadResult result, HSSettings settingsContext)
    {
        Debug.Assert(result is not null);
        Debug.Assert(!string.IsNullOrWhiteSpace(result.Url));

        switch (settingsContext.ActionAfterUpload)
        {
            case UploadHandlingAction.Flyout:
                (new UploadResultForm(result, settingsContext)).Show();
                break;
            case UploadHandlingAction.CopyToClipboard:
                {
                    var success = ClipboardEx.SetText(result.Url);
                    if (settingsContext.ShowCopyConfirmation)
                    {
                        if (success)
                            NotificationManager.ShowCopyConfirmation(result.Url);
                        else
                            NotificationManager.CopyingFailed(result.Url);
                    }
                }
                break;
            case UploadHandlingAction.None:
                // Intentionally left empty
                break;
            default:
                // Intentionally left empty
                break;
        }
    }
}
